Constraint-preserving BCs for GH

prayush opened the pull request:
## Proposed changes

We add here the infrastructure (an `Action`) to impose Bjorhus-type boundary conditions for the generalized harmonic evolution system. This action sets desired values for the characteristic projected time derivatives of the evolved fields on the outer domain boundary. In addition to the action, we also implements here the actual conditions imposed by `SpEC` on the outer boundary for GeneralizedHarmonic. These include:
 - constraint preserving boundary conditions on `dt<VPsi>`, `dt<VZero>` and `dt<VMinus>`,
 - physical boundary conditions on `dt<VMinus>`, which restrict its two physical degrees of freedom,
 - gauge-dependent boundary conditions on `dt<VMinus>`, which restrict the gauge d.o.f.

**Update 10/06/2020**:

This PR has been updated following the first round of review by @geoffrey4444 . Since the changes were major and the fundamental design was updated, all fixups have been squashed and the updated code been refactored into multiple commits. The first 3 commits are now also independent PRs #2527 #2528 and #2529 . Please review these 3 commits over there, so we can merge them first and separately.


### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 commented:
@prayush please rebase? That might fix clang-tidy. Do you know why the tests fail?

prayush commented:
@geoffrey4444 rebased on develop, including recently merged GH-related PRs. Hoping that fixes the clang-tidy failures. Tests were failing because the GH executable's `execute` test timed out. I have also increased the allowed range from 8s to 12s, hoping that would avoid timeouts.

geoffrey4444 reviewed (changes requested):
Test still times out...any way to speed it up?

geoffrey4444 commented on src/Evolution/Initialization/DgDomain.hpp:
```diff
@@ -99,6 +99,7 @@ struct Domain {
     using simple_tags = db::AddSimpleTags<
         ::domain::Tags::Mesh<Dim>, ::domain::Tags::Element<Dim>,
         ::domain::Tags::ElementMap<Dim, Frame::Grid>,
+        ::domain::Tags::ElementMap<Dim, Frame::Inertial>,
```
I'm confused by this commit. If you need this to compute characteristics, then how does the moving mesh code already in develop work correctly?

geoffrey4444 replied:
Discussion on call 5/18: we don't see why this is needed

prayush replied:
oops, sorry this crept in from the development branch where I had included it for the scalar wave system (which needs this commit).

geoffrey4444 commented on src/Evolution/Initialization/DgDomain.hpp:
```diff
@@ -142,6 +143,14 @@ struct Domain {
+    ElementMap<Dim, Frame::Inertial> element_map_inertial{
+        element_id, my_block.is_time_dependent()
+                        // In case of a time dependent domain, this
+                        // element map should / would not be used
```
I don't understand this comment...can you please clarify? Does this code work correctly with moving mesh and time-dependent maps? Does it work with stationary maps as well? Or both?

prayush replied:
removed commit

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:
```diff
@@ -948,6 +948,36 @@ struct GaugeHImplicitFrom3p1QuantitiesCompute : GaugeH<SpatialDim, Frame>,
   using base = GaugeH<SpatialDim, Frame>;
 };
 
+/*!
```
Do you need this for observers? If not, is it possible to implement this as only a free function, so that we can avoid adding compute tags?

prayush replied:
the spatial derivative of GaugeH is needed by `FConstraint` and `TwoIndexConstraint`. To observe them I need to have `::Tags::deriv<GaugeH<SpatialDim, Frame>, tmpl::size_t<SpatialDim>, Frame>` in the databox.

I can make this a free function, and modify the compute tags for the F- and 2index-constraints to take in the spacetime derivative of GaugeH (which is already available), and extract the spatial deriv from it before continuing the computation of constraints. The downside of this is only that the computation will be repeated (once for each constraint).  Is this something you'd prefer?

**update:** Implemented the change suggested in the previous paragraph.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -56,19 +56,29 @@ struct InitializeConstraints {
                     const ParallelComponent* const /*meta*/) noexcept {
     using compute_tags = tmpl::flatten<db::AddComputeTags<
         GeneralizedHarmonic::Tags::GaugeConstraintCompute<Dim, frame>,
+        // GeneralizedHarmonic::Tags::ThreeIndexConstraintCompute<Dim, frame>,
```
Why comment out the three-index constraint? Also, why are these constraints necessary for this PR?

prayush replied:
They are not. I had added all constraints to the executable for (mostly) testing and validating purposes. I agree that maybe this change should be made outside of this PR. This also ties into <https://github.com/sxs-collaboration/spectre/pull/1558#discussion_r444492000>  which is only needed because of newly included constraints in the executable. 

I'll make this a separate PR, which would (a) change the dependencies of `FConstraint` and `TwoIndexConstraint` as discussed [above](https://github.com/sxs-collaboration/spectre/pull/1558#discussion_r444492000), and (b) include more constraint observers in the executable.

geoffrey4444 reviewed (changes requested):
Took a look at the first 4 commits

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsHelpers.hpp:
```diff
@@ -0,0 +1,316 @@
+namespace BoundaryConditions_detail {
+template <size_t VolumeDim>
+double min_characteristic_speed(
+    const typename GeneralizedHarmonic::Tags::CharacteristicSpeeds<
```
Why not just explicitly put the type in here?

prayush replied:
changed

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsHelpers.hpp:
```diff
@@ -0,0 +1,316 @@
+namespace Actions {
+namespace BoundaryConditions_detail {
+template <size_t VolumeDim>
+double min_characteristic_speed(
```
Because the functions in this file are visible elsewhere, being in an hpp file that others might include, please add some doxygen documentation for them

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsHelpers.hpp:
```diff
@@ -0,0 +1,316 @@
+}  // namespace BoundaryConditions_detail
+}  // namespace Actions
+
+// Eq. (2.20) of Kidder, Scheel & Teukolsky (KST) [gr-qc/0105031v1]
```
Why do you need these "KST" functions, instead of just using the existing functions? And why are they here, instead of in PointwiseFunctions/GeneralRelativity?

prayush replied:
There is no real need to get the intermediate KST variables, as you point out. I originally did not have them, but during debugging vs SpEC I was comparing all intermediates as well, and thats how these functions came about. 

I have removed the need for all KST variables now (as it should be). We still needed an overload of the `ricci_tensor` function that computes the tensor from GH variables `Phi` and its numerical spatial derivative, instead of from the 1st and 2nd-order numerical spatial derivative of spacetime metric.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+                 typename Metavariables::boundary_condition_tag>;
+
+ private:
+  /* ------------------------------------------------------------------------
```
Remove all the dashes and extra lines...pretty sure these go against are style guide

prayush replied:
removed

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+          domain::Tags::BoundaryDirectionsInterior<VolumeDim>,
+          ::Tags::Normalized<domain::Tags::UnnormalizedFaceNormal<
+              VolumeDim, Frame::Inertial>>>>(box);
+      // const auto& external_bdry_vars = db::get<domain::Tags::Interface<
```
please remove commented out lines that aren't needed

prayush replied:
removed, sorry these shouldn't have been here.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+      // variables_tag>>(box);
+      const auto& volume_all_vars = db::get<variables_tag>(box);
+      const auto& volume_all_dt_vars = db::get<dt_variables_tag>(box);
+      const auto& external_bdry_char_speeds = db::get<domain::Tags::Interface<
```
Can you instead get the ingredients from the DataBox but compute the char speeds with the free function directly?

prayush replied:
Yes that is easy to change, but may I ask why? The primary purpose of having char speeds' compute tags in the data box is for their use in boundary conditions (internal / external). Why would we not want to use them, unless the plan is to eliminate the compute tag for char speeds completely?

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+            make_not_null(&buffer), box, direction, dimension, mesh, vars,
+            dt_vars, unit_normal_one_form, char_speeds);
+
+        // FIXME: Are there incoming char speeds on the inner boundary?
```
Is the intention here to add a check for incoming char speeds on the inner boundary and error out if they are found? If so, how will this be implemented, and why not implement this as part of this PR?

prayush replied:
This was intended for debugging only, sorry I should have removed it. The test used here to determine if an interface is inside of a horizon is rudimentary. A proper implementation primarily requires a mechanism to determine all external interfaces in the domain that are completely inside of any horizon. As an inner boundary is likely to remain an inner boundary throughout the evolution, and because classifying all inner boundaries as such is a global operation, this is best done once at the start of the evolution. It can be done either

1. dynamically once after the first time all apparent horizons have been found. or
1. statically by manually tagging all inner interfaces as such during initialization (using perhaps the initial guesses for Ahs).

Going with (2) also adds the complexity of inheriting the "internal boundary" tag appropriately when h-refining the domain. Beyond this, changes needed to add this check to the BC action introduced in this PR will be trivial. There already is a check for boundaries with only outgoing characteristics, this would be an additional if statement checking for the same to be false when the interface is an inner boundary.

Doing either of (1) or (2) requires
- changes to the apparent horizon finding code (where the trigger for tagging inner interfaces must lie), 
- some mechanism to tag 'inner' interfaces as such (perhaps inside the `Interface` tag, this part is not very clear to me yet), and
- additional code to actually do the tagging - whatever be its trigger.

These changes, while small, are best kept separate from this PR as they involve structural changes mostly outside of GH. Especially so since @nilsdeppe is generalizing the whole BC imposition framework, this check might be better placed over there as it is not specific to GH at all. Thoughts?

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+          const auto& x =
+              get<::Tags::TempI<37, VolumeDim, Frame::Inertial, DataVector>>(
+                  buffer);
+          // If radius of surface is less than 10, then assume we are on an
```
Below it looks like the minimum is 2, which is insufficient for BBHs. Even 10 might be insufficient. Probably this should be a parameter specified in the input file, unless @nilsdeppe has a way to identify whether a surface is on the interior or not

prayush replied:
see [above](https://github.com/sxs-collaboration/spectre/pull/1558#discussion_r454817202)

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+                    char_speeds);
+            if (min_speed <= 0.0) {
+              Parallel::printf(
+                  "\nWARNING: Incoming char speeds on INNER boundary at t=%f\n",
```
Don't just print this, stop the simulation with a runtime error

prayush replied:
see [above](https://github.com/sxs-collaboration/spectre/pull/1558#discussion_r454817202)

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+              Parallel::printf(
+                  "  Min speed %f at min (inertial) radius %f (%f)\n\n",
+                  min_speed, sqrt(min_r_squared), sqrt(min_inertial_r_squared));
+              // Is the face normal outward facing, really?
```
what does "really" mean here?

prayush replied:
see [above](https://github.com/sxs-collaboration/spectre/pull/1558#discussion_r454817202)

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+                  "  Min speed %f at min (inertial) radius %f (%f)\n\n",
+                  min_speed, sqrt(min_r_squared), sqrt(min_inertial_r_squared));
+              // Is the face normal outward facing, really?
+              for (size_t i = 0; i < 5; ++i) {
```
Add a comment that these are diagnostics to help diagnose incoming char speeds (or a comment explaining what they're for, if that's not it)

prayush replied:
see [above](https://github.com/sxs-collaboration/spectre/pull/1558#discussion_r454817202)

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+              // Get desired values of CharProjection<dt<U>>
+              //
+              // At all points on the interface where the char speed of any
+              // (given) characteristic field is +ve, we "do nothing", and
```
spell out positive and negative please

prayush replied:
done

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+              // `set_bc_when_char_speed_is_negative`.
+              const auto bc_dt_u_psi =
+                  BoundaryConditions_detail::set_bc_when_char_speed_is_negative(
+                      get<::Tags::Tempaa<22, VolumeDim, Frame::Inertial,
```
all these temporaries are hard for me to follow. Can you introduce aliases above and use those, so they have sensible names?

prayush replied:
Rewrote the handling of temporaries. See [below](https://github.com/sxs-collaboration/spectre/pull/1558#discussion_r456866417).

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+                      unit_normal_one_form);
+              // Now store final values of dt<U> in suitable data structure
+              // FIXME: How can I extract this list of dt<U> tags directly
+              // from `dt_variables_tag`?
```
@nilsdeppe do you know the answer to this fix me question?

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions.hpp:
```diff
@@ -0,0 +1,413 @@
+                      // BC choice for U_0
+                      VZeroBcMethod::ConstraintPreservingBjorhus,
+                      // BC choice for U_+
+                      VPlusBcMethod::Freezing,
```
Add a comment why this one is freezing (is it because it's char speeds are never incoming?)

prayush replied:
added

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp:
```diff
@@ -0,0 +1,1447 @@
+};
+
+template <size_t VolumeDim>
+using all_local_vars = tmpl::list<
```
can you please give these some aliases with sensible names, so we can refer to these by names later on? this is really hard to follow

geoffrey4444 replied:
Update: maybe a comment saying you'll introduce aliases below for each quantity would help the reader

prayush replied:
Overhauled the computation of intermediate variables. Now, the computation and storage are both handled by a single standalone computer class and we use human-readable aliases everywhere instead of indexed names.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp:
```diff
@@ -0,0 +1,1447 @@
+                    tmpl::size_t<VolumeDim>, Frame::Inertial>,
+      ::Tags::deriv<Tags::Phi<VolumeDim, Frame::Inertial>,
+                    tmpl::size_t<VolumeDim>, Frame::Inertial>,
+      // ---- constraint damping parameters
```
Please don't leave commented-out lines in a PR that is intended to be merged into develop. Instead, if you want to remove something the reader would expect to be there, a comment explaining why it's not there would be appropriate

prayush replied:
Removed these, sorry!

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp:
```diff
@@ -0,0 +1,1447 @@
+      const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>& phi) noexcept;
+};
+
+template <typename ReturnType, size_t VolumeDim>
```
I think each of these needs doxygen help with the equations being implemented, with a citation to the reference and equation number where it comes from

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp:
```diff
@@ -0,0 +1,1447 @@
+    const std::array<DataVector, 4>& char_speeds) noexcept {
+  ASSERT(get_size(get<0, 0>(*bc_dt_u_psi)) ==
+             get_size(get<0>(unit_interface_normal_vector)),
+         "Size of input variables and temporary memory do not match.");
```
Print both sizes as well, instead of just saying they don't match

prayush replied:
included sizes in this and all other asserts

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp:
```diff
@@ -0,0 +1,1447 @@
+         "Size of input variables and temporary memory do not match.");
+  for (size_t a = 0; a <= VolumeDim; ++a) {
+    for (size_t b = 0; b <= VolumeDim; ++b) {
+      // For a chosen (a, b):
```
comments like this should be doxygen comments so we can check the typesetting etc.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp:
```diff
@@ -0,0 +1,1447 @@
+    }
+
+    // ComputeRicciFromD(kst_D, d_kst_D, Invg, Ricci3);
+    auto ricci_3 = GeneralizedHarmonic::spatial_ricci_tensor_from_KST_vars(
```
why can't you use the existing pointwise functions to get the 3-ricci and christoffel second kind?

prayush replied:
Changed these to use existing christoffel functions, but I had to add an overload of the `ricci_tensor` function anyway as the existing version (there's only one) needs as input quantities that haven't been computed here yet.

geoffrey4444 reviewed (changes requested):
Took a pass through the code but not yet the test

prayush commented:
Hi @geoffrey4444 - most of the code in this PR has now been fixed per your comments above. Most important changes were (A) the computation of temporaries, (B) removal of auxiliary code not necessary for this PR's purpose, and (C) a few bugfixes. Could you please have a look and let me know when I can squash the fixups. (the merge conflicts after straightforward to fix after squashing)? Thanks!

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.hpp:
```diff
@@ -189,153 +177,87 @@ struct ImposeBjorhusBoundaryConditions {
-              // FIXME: How can I extract this list of dt<U> tags directly
-              // from `dt_variables_tag`?
+              // FIXME: Can we extract this list of dt<U> tags directly from
+              // `dt_variables_tag`?
```
@nilsdeppe ?

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.hpp:
```diff
@@ -391,18 +312,21 @@ struct ImposeBjorhusBoundaryConditions {
-    // setting BCs on individual characteristic variables
+    // One can insert logic here that determines / selects from various options
+    // for setting BCs on individual characteristic variables.
+    // `ConstraintPreservingBjorhus` are hardcoded for now.
```
Why not ConstraintPreservingPhysicalBjorhus?

prayush replied:
This has changed in the recent version of this file, now there is no default. The BC choice is governed by the options passed to `Metavars` during compilation, as a result we have different versions of `EvolveGeneralizedHarmonic` available for each possible BC choice (see [here](https://github.com/sxs-collaboration/spectre/blob/180995eb56b8c558d614a6890c451824d9ce3c92/src/Evolution/Executables/GeneralizedHarmonic/CMakeLists.txt#L59))

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/BjorhusHelpers.hpp:
```diff
@@ -82,85 +81,102 @@ T set_bc_when_char_speed_is_negative(const T& rhs_char_dt_u,
-tnsr::ii<DataType, VolumeDim, Frame> spatial_ricci_tensor_from_KST_vars(
-    const tnsr::ijj<DataType, VolumeDim, Frame>& kst_var_D,
-    const tnsr::ijkk<DataType, VolumeDim, Frame>& d_kst_var_D,
+tnsr::ii<DataType, VolumeDim, Frame> ricci_tensor(
```
Would anyone besides this BC want to get Ricci in this way? If so, probably it should be added with a unit test to PointwiseFunctions/GeneralRelativity/Ricci.?pp?

prayush replied:
Not sure, but I like the idea of moving this to Pointwisefunctions. Since the inputs to this method are specific to GH, I've moved it to `PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.?pp?`.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/BjorhusImpl.hpp:
```diff
@@ -0,0 +1,1327 @@
+    const auto& shift =
+        get<gr::Tags::Shift<VolumeDim, Frame::Inertial, DataVector>>(
+            vars_on_this_slice);
+    // const auto& spatial_metric =
```
If you're not using this, please remove the commented-out lines

prayush replied:
removed

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/BjorhusImpl.hpp:
```diff
@@ -0,0 +1,1327 @@
+                          const Variables<DtVarsTagsList>& /* dt_vars */,
+                          const tnsr::i<DataVector, VolumeDim, Frame::Inertial>&
+                          /* unit_normal_one_form */) noexcept {
+    // Not using auto below to enforce a loose test on the quantity being
```
You are using auto below, why do you say you are not?

prayush replied:
oops, relic comment, now removed.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:
```diff
@@ -100,10 +100,10 @@ void two_index_constraint_add_term_4_of_11(
 void two_index_constraint_add_term_5_of_11(
     const gsl::not_null<tnsr::ia<DataType, SpatialDim, Frame>*> constraint,
-    const tnsr::ia<DataType, SpatialDim, Frame>& d_gauge_function) noexcept {
+    const tnsr::ab<DataType, SpatialDim, Frame>& d_gauge_function) noexcept {
```
Why this change? d_gauge_function is the spatial derivative

prayush replied:
because the spacetime deriv of gauge_function is already available in the databox, instead of extracting the spatial deriv elsewhere, we just now do that in here itself.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/BjorhusImpl.hpp:
```diff
@@ -0,0 +1,1327 @@
+// Distributed under the MIT License.
```
This file is really long...any way to factor it?

prayush replied:
Reduced file length by 35% by moving the intermediates' computation to separate file. Refactoring beyond this will require splitting variable-wise, which does not seem very reasonable to me. Any other ideas are welcome!

geoffrey4444 reviewed (changes requested):
Thanks for doing this, @prayush! I have some comments, but I think soon it will be ready for someone else to take a look.

One big comment I have is that the PR is very large, and at least one of the files is very long. It would be much easier to review if it could be factored somehow into smaller files and multiple commits or multiple PRs

prayush commented:
Hi @geoffrey4444 - Addressed your specific comments above. Before refactoring the whole PR into smaller bits, would you be okay with me squashing the fixups and rebasing with develop?

> Thanks for doing this, @prayush! I have some comments, but I think soon it will be ready for someone else to take a look.
> 
> One big comment I have is that the PR is very large, and at least one of the files is very long. It would be much easier to review if it could be factored somehow into smaller files and multiple commits or multiple PRs

geoffrey4444 commented:
I'm ok to squash if no one else objects

geoffrey4444 commented:
No one objected for 18 days, so o guess you should squash

prayush committed:
Add aliases for temp tensors Ijj, ijk, and ijaa

prayush committed:
Simplify input arguments for the two-index and F-constraints

Both of these require the spatial derivative of the gauge
source function H. As the spacetime derivative of H is almost
always added to the evolution databox anyway, it is simpler
for these constraints' computing methods and compute tags
to directly accept the spacetime derivative of H as input
and extract the spatial derivative internally. This commit
makes this simplification.

prayush committed:
Add method to compute the spatial Ricci tensor from GR and GH vars

Spatial Ricci tensor can be computed from the spatial metric, GH var
Phi and its spatial derivative [Kidder, Scheel and Teukolsky (2001)].
This is subsequently needed to compute the propagating mode of Weyl
tensor, needed to impose boundary conditions on incoming physical
modes at external boundaries. This commit adds functions for the same.

prayush committed:
Add Bjorhus-type constraint preserving and physical boundary conditions for GH

prayush committed:
Add compute tags to GH initialization that are needed for Bjorhus BCs

prayush committed:
Add Gh executables that use Bjorhus-type BCs on external faces

markscheel commented:
@prayush I'm a bit confused. Doesn't this PR need to wait until the dependent ones are merged?  Did you want me to review some of those instead of this one?

prayush commented:
> @prayush I'm a bit confused. Doesn't this PR need to wait until the dependent ones are merged? Did you want me to review some of those instead of this one?

Hi @markscheel - Yes it would be good to do those first (thanks for your review on #2528 already). Those two (#2528 and #2529) are short and relatively straightforward however, while this one is much larger (by a factor of 10), and so I primarily wanted to get someone to volunteer for this first while the dependencies get reviewed in parallel at their own pace :smile:
